internal/pkg/consensus/validation.go

```diff
@@ -47,54 +47,41 @@ func init() {
 	errNonceTooHighCt = metrics.NewInt64Counter("consensus/msg_nonce_high_err", "Number of messages with nonce too high")
 }
 
-// DefaultMessageValidator validates incoming signed messages.
-type DefaultMessageValidator struct {
-	allowHighNonce bool
+// MessageSelectionChecker checks for miner penalties on signed messages
+type MessagePenaltyChecker struct {
+	api penaltyCheckerAPI
 }
 
-// NewDefaultMessageValidator creates a new default validator.
-// A default validator checks for both permanent semantic problems (e.g. invalid signature)
-// as well as temporary conditions which may change (e.g. actor can't cover gas limit).
-func NewDefaultMessageValidator() *DefaultMessageValidator {
-	return &DefaultMessageValidator{}
+// penaltyCheckerAPI allows the validator to access latest state
+type penaltyCheckerAPI interface {
+	Head() block.TipSetKey
+	GetActorAt(ctx context.Context, tipKey block.TipSetKey, addr address.Address) (*actor.Actor, error)
 }
 
-// NewOutboundMessageValidator creates a new default validator for outbound messages. This
-// validator matches the default behaviour but allows nonces higher than the actor's current nonce
-// (allowing multiple messages to enter the mpool at once).
-func NewOutboundMessageValidator() *DefaultMessageValidator {
-	return &DefaultMessageValidator{allowHighNonce: true}
+func NewMessagePenaltyChecker(api penaltyCheckerAPI) *MessagePenaltyChecker {
+	return &MessagePenaltyChecker{
+		api: api,
+	}
 }
 
-// Validate checks that a message is semantically valid for processing, returning any
-// invalidity as an error.
-func (v *DefaultMessageValidator) Validate(ctx context.Context, msg *types.UnsignedMessage, fromActor *actor.Actor) error {
-	if msg.From == msg.To {
-		return errSelfSend
+// PenaltyCheck checks that a message is semantically valid for processing without
+// causing miner penality.  It treats any miner penalty condtion as an error.
+func (v *MessagePenaltyChecker) PenaltyCheck(ctx context.Context, msg *types.UnsignedMessage) error {
```

Comment:
`error` is no longer a good return type to use here, IMO. It's not an error in any of the usual senses, just a signal that this message is not appropriate for this state (but may be fine later). I would say use bool, but it's handy to get a reason string too. Maybe (bool, string)? Your call if you want to stay with error, but please document that it's a misuse of error.

Reply:
Not sure I agree but I've file an issue anyway for getting this part of the system working and we can work it out there: #3873